scipy/stats/_hypotests.py

```diff
@@ -663,3 +663,274 @@ def somersd(x, y=None):
+        group (80%) eventually became infected with influenza whereas for the
+        vaccine group, only 7 of the 15 subjects (47%) became infected. The
+        data are tabulated as a 2 x 2 table::
+
+                Vaccine  Placebo
+            Yes     7        12
+            No      8        3
+                    15        15
+
+    When working with statistical hypothesis testing, we usually use a
+    threshold probability or a significance level upon which we decide
+    to reject or not the null hypothesis :math:`H_0`. A commonly used
+    significance level is 5%. In this example, we are using the
+    `alternative` parameter with the "less" option, because the vaccine has
+    either no effect (:math:`H_0` hypothesis) or a positive effect
+    (:math:`H_1` hypothesis). Therefore, under the null hypothesis
+    (:math:`H_0`), the probability :math:`p_2` of catching the virus
+    **without** any vaccine, is either equal or lower than :math:`p_1`,
+    the probability of catching the virus **with** the vaccine.
+    To compute the p-value with Banard test, let's call `barnard_exact` as
+    follow :
+
+    >>> import scipy.stats as stats
+    >>> res = stats.barnard_exact([[7, 12], [8, 3]], alternative="less")
+    >>> res.statistic
+    1.894...
+    >>> res.pvalue
+    0.03407...
+
+    Then, the probability of obtaining test results at least as extreme as the
+    data observed above is around 3.4%. Since our p-value is under our
+    significance level defined above, we reject :math:`H_0`. The vaccine has
+    a positive effect and we have :math:`p_1 \leq p_2`; the probability of
+    contracting the virus if we have been vaccinated is lower than if
+    we are not.
+
+    We can compare with Fisher's exact test which produces an exact
+    p-value of 6.4% :
+
+    >>> _, pvalue = stats.fisher_exact([[7, 12], [8, 3]], alternative="less")
+    >>> pvalue
+    0.0640...
+
+    With the same threshold probability of 5%, `fisher_exact` is accepting
+    :math:`H_0` while `barnard_exact` is rejecting it. As stated in [1]_,
+    Barnard's test is uniformly more powerful than Fisher's exact test
+    because Barnard's test do not condition on any margin. Fisher's test
+    should only be used when both sets of marginals are fixed.
+
+    """
+    if num_it <= 0:
+        raise ValueError(
+            "Number of iterations `num_it` must be strictly positive, "
+            f"found {num_it!r}"
+        )
+
+    c = np.asarray(table, dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0):
+        # If both values in column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return BarnardExactResult(np.nan, 1.0)
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
```

Comment:
Yes that's true, this is used latter indeed in the `_binomial_maximisation_of_p_value_with_nuisance_param`, but since I redefined n, I can remove this line